api/utils/web_scraping.py

```diff
@@ -108,35 +110,89 @@ def get_teachers(self, data: list) -> list:
 
         if len(teachers) == 0:
             teachers.append("A definir")
-        
+
         return teachers
 
-    def get_schedules(self, data: str) -> list:
+    def get_schedules_and_intervals(self, data: str) -> list[list[str], list[tuple[int, int]]]:
         regex = "\d+[MTN]\d+"
-        occurrences = findall(regex, data)
-        
-        return occurrences
-    
-    def get_special_dates(self, data: str) -> list:
+        occurrences = finditer(regex, data)
+        values = [[], []]
+
+        for value in occurrences:
+            values[0].append(value.group())
+            values[1].append((value.start(), value.end()))
+
+        return values
+
+    def check_start(self, *args, **kwargs) -> bool:
+        start_index = kwargs.get("start_index")
+        last_included = kwargs.get("last_included")
+
+        end_interval = kwargs.get("interval")[1]
+        already_included = kwargs["index"] + 1 > last_included
+        value_start_check = kwargs.get("value").start() > end_interval
+
+        return start_index is None and value_start_check and already_included
+
+    def check_end(self, *args, **kwargs) -> bool:
+        end_index = kwargs.get("end_index")
+
+        start_interval = kwargs.get("interval")[0]
+        value_start_check = kwargs.get("value").start() < start_interval
+
+        return end_index is None and value_start_check
+
+    def get_start_index(self, intervals, last_included, value):
+        for index, interval in enumerate(intervals):
+            if self.check_start(start_index=index, last_included=last_included, interval=interval, index=index, value=value):
+                return index + 1
+        return None
+
+    def get_end_index(self, intervals, value):
+        for index, interval in enumerate(intervals):
```

Comment:
Necessário adicionar tipo de retorno da função.
```suggestion
    def get_end_index(self, intervals, value) -> Optional[int]:
```